controllers/rcj_soccer_referee_supervisor/referee/referee.py

```diff
@@ -1,44 +1,248 @@
+
+        self.event_messages_to_draw.append((self.time, message))
+
+    def process_and_draw_event_messages(self):
+        """Process and draw event messages from the queue"""
+        messages = []
+        for time, msg in self.event_messages_to_draw:
+            messages.append("{} - {}".format(time_to_string(time), msg))
+
+        self.sv.draw_event_messages(messages)
+
+    def reset_checkers(self, object_name: str):
+        """Reset rule checkers for the specified object.
+
+        Args:
+            object_name (str): Either "ball" or the robot's name.
+        """
+        self.progress_check[object_name].reset()
+        if object_name != "ball":
+            self.penalty_area_check[object_name].reset()
+
+    def reset_ball_position(self):
+        """Reset the position of the ball."""
+        self.sv.set_ball_position(BALL_INITIAL_TRANSLATION)
+        self.ball_stop = 2
+        self.reset_checkers("ball")
+
+    def reset_robot_position(self, robot_name: str):
+        """Reset robot's position to the initial one.
+
+        Args:
+            robot_name (str): The robot to reset the position for
+        """
+        self.sv.reset_robot_velocity(robot_name)
+
+        translation = ROBOT_INITIAL_TRANSLATION[robot_name].copy()
+        translation = self._add_initial_position_noise(translation)
+
+        self.sv.set_robot_position(robot_name, translation)
+        self.sv.set_robot_rotation(
+            robot_name, ROBOT_INITIAL_ROTATION[robot_name]
+        )
+
+        self.reset_checkers(robot_name)
+
+    def reset_positions(self):
+        """
+        Reset the positions of the ball as well as the robots to the initial
+        position.
+        """
+        self.reset_ball_position()
+
+        # reset the robot positions
+        for robot in ROBOT_NAMES:
+            self.reset_robot_position(robot)
 
+    def reset_team_for_kickoff(self, team: str) -> str:
+        """
+        Given a team name ('B' or 'Y'), set the position of the third robot on
+        the team to "kick off" (inside the center circle).
+
+        Args:
+            team (str): 'B' for blue or 'Y' for yellow team
+
+        Returns:
+            str: Name of the robot that is kicking off.
+        """
+        # Always kickoff with the third robot
+        robot = f"{team}3"
+
+        self.sv.set_robot_position(robot, KICKOFF_TRANSLATION[team])
+        self.sv.set_robot_rotation(robot, ROBOT_INITIAL_ROTATION[robot])
+
+        return robot
 
-class RCJSoccerReferee(RCJSoccerSupervisor):
     def check_robots_in_penalty_area(self):
         """
         Check whether robots are violating rule not to stay in
         penalty area for longer period of time
         """
         for robot in ROBOT_NAMES:
-            pos = self.robot_translation[robot]
-            self.penalty_area_chck[robot].track(pos, self.time)
+            pos = self.sv.get_robot_translation(robot)
+            self.penalty_area_check[robot].track(pos, self.time)
```

Comment:
@Adman I guess I should say "Thank you" here :smile:

Reply:
Why? :)